### PR TITLE
Reputation log

### PR DESCRIPTION
**Pull request by RainbowDashLabs**
This PR introduces the log command to see the latest received and donated reputations per user and reputation for messages.

### CODE REVIEW

File: src/main/java/de/chojo/repbot/data/wrapper/ReputationLogEntry.java [hunk 1]
```diff
@@ -7,11 +7,25 @@
 @Getter
 @AllArgsConstructor
 public class ReputationLogEntry {
+    private static final String PATH = "https://discord.com/channels/%s/%s/%s";
+
     private final long guildId;
     private final long channel_id;
```
Comment (yannicklamprecht) [inline src/main/java/de/chojo/repbot/data/wrapper/ReputationLogEntry.java:13]:
```suggestion
    private final long channelId;
```

File: src/main/java/de/chojo/repbot/data/wrapper/ReputationLogEntry.java [hunk 2]
```diff
@@ -7,11 +7,25 @@
...
     private final long donorId;
     private final long receiverId;
     private final long messageId;
-    private final long ref_message_id;
+    private final long refMessageId;
     private final ThankType type;
+
+    public String getMessageJumpLink() {
+        return String.format(PATH, guildId, channel_id, messageId);
```
Comment (yannicklamprecht) [inline src/main/java/de/chojo/repbot/data/wrapper/ReputationLogEntry.java:21]:
```suggestion
        return String.format(PATH, guildId, channelId, messageId);
```

File: src/main/java/de/chojo/repbot/data/wrapper/ReputationLogEntry.java [hunk 3]
```diff
@@ -7,11 +7,25 @@
...
+    }
+
+    public String getRedMessageJumpLink() {
+        return String.format(PATH, guildId, channel_id, refMessageId);
```
Comment (yannicklamprecht) [inline src/main/java/de/chojo/repbot/data/wrapper/ReputationLogEntry.java:25]:
```suggestion
        return String.format(PATH, guildId, channelId, refMessageId);
```

File: src/main/java/de/chojo/repbot/data/ReputationData.java [hunk 4]
```diff
@@ -1,197 +1,296 @@
...
+                    donor_id,
+                    receiver_id,
+                    message_id,
+                    received,
+                    ref_message_id,
+                    channel_id,
+                    cause
+                FROM
+                    reputation_log
+                where
```
Comment (yannicklamprecht) [inline src/main/java/de/chojo/repbot/data/ReputationData.java:200]: #fucke-up-line-endings before?

File: src/main/java/de/chojo/repbot/commands/Log.java [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+        if ("donated".equalsIgnoreCase(cmd)) {
+            return donated(eventWrapper, context.subContext(cmd));
+        }
+        if ("message".equalsIgnoreCase(cmd)) {
+            return message(eventWrapper, context.subContext(cmd));
+        }
+        return false;
+    }
+
+    private boolean message(MessageEventWrapper eventWrapper, CommandContext subContext) {
```
Comment (yannicklamprecht) [inline src/main/java/de/chojo/repbot/commands/Log.java:53]: Duplicate code.

File: src/main/java/de/chojo/repbot/commands/Log.java [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+        var message = new LocalizedEmbedBuilder(eventWrapper)
+                .setAuthor(eventWrapper.localize("command.log.messageLog", Replacement.create("ID", optMessageId.get())))
+                .setDescription(log)
+                .build();
+        eventWrapper.reply(message).queue();
+
+        return true;
+    }
+
+    private boolean donated(MessageEventWrapper eventWrapper, CommandContext context) {
```
Comment (yannicklamprecht) [inline src/main/java/de/chojo/repbot/commands/Log.java:76]: Duplicate code.

File: src/main/java/de/chojo/repbot/commands/Log.java [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+        var message = new LocalizedEmbedBuilder(eventWrapper)
+                .setAuthor(eventWrapper.localize("command.log.donatedLog", Replacement.create("USER", user.getAsTag())),
+                        user.getAvatarUrl())
+                .setDescription(log)
+                .build();
+        eventWrapper.reply(message).queue();
+        return true;
+    }
+
+    private boolean received(MessageEventWrapper eventWrapper, CommandContext context) {
```
Comment (yannicklamprecht) [inline src/main/java/de/chojo/repbot/commands/Log.java:101]: Duplicate code.

**Review by yannicklamprecht — CHANGES_REQUESTED** [review]

File: src/main/java/de/chojo/repbot/data/ReputationData.java (hunk 4 above)
Comment (RainbowDashLabs) [reply src/main/java/de/chojo/repbot/data/ReputationData.java:200]: probably yes... Really need linux...

File: src/main/java/de/chojo/repbot/commands/Log.java [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+                .setDescription(log)
+                .build();
+        eventWrapper.reply(message).queue();
+
+    }
+
+    private boolean donated(MessageEventWrapper eventWrapper, CommandContext context, User user) {
+        var limit = context.argInt(1).orElse(10);
+
+        var userDonatedLog = reputationData.getUserDonatedLog(user, eventWrapper.getGuild(), Math.max(5, Math.min(limit, 50)));
```
Comment (yannicklamprecht) [inline src/main/java/de/chojo/repbot/commands/Log.java:94]: Code probably can be further reduced by using a Supplier and a TriFunction<User, Guild, int, T>
Not 100% needed, just an idea.

**Review by yannicklamprecht — APPROVED** [review]

**Pull request merged**